mobile/library/jni/jni_impl.cc

```diff
@@ -827,24 +824,58 @@ extern "C" JNIEXPORT jlong JNICALL Java_io_envoyproxy_envoymobile_engine_JniLibr
+                                         bool end_stream, envoy_stream_intel stream_intel) {
+        envoy_headers bridge_headers = Envoy::Http::Utility::toBridgeHeaders(headers);
+        jvm_on_response_headers(bridge_headers, end_stream, stream_intel,
+                                java_stream_callbacks_global_ref);
+      };
+  stream_callbacks.on_data_ = [java_stream_callbacks_global_ref](
+                                  const Envoy::Buffer::Instance& buffer, uint64_t length,
+                                  bool end_stream, envoy_stream_intel stream_intel) {
+    envoy_data bridge_data = Envoy::Data::Utility::toBridgeData(buffer, length);
+    jvm_on_response_data(bridge_data, end_stream, stream_intel, java_stream_callbacks_global_ref);
+  };
+  stream_callbacks.on_trailers_ =
+      [java_stream_callbacks_global_ref](const Envoy::Http::ResponseTrailerMap& trailers,
+                                         envoy_stream_intel stream_intel) {
+        envoy_headers bridge_trailers = Envoy::Http::Utility::toBridgeHeaders(trailers);
+        jvm_on_response_trailers(bridge_trailers, stream_intel, java_stream_callbacks_global_ref);
+      };
+  stream_callbacks.on_complete_ =
+      [java_stream_callbacks_global_ref](envoy_stream_intel stream_intel,
+                                         envoy_final_stream_intel final_stream_intel) {
+        jvm_on_complete(stream_intel, final_stream_intel, java_stream_callbacks_global_ref);
+      };
+  stream_callbacks.on_error_ = [java_stream_callbacks_global_ref](
+                                   Envoy::EnvoyError error, envoy_stream_intel stream_intel,
+                                   envoy_final_stream_intel final_stream_intel) {
+    envoy_error bridge_error = Envoy::Bridge::Utility::toBridgeError(error);
+    jvm_on_error(bridge_error, stream_intel, final_stream_intel, java_stream_callbacks_global_ref);
+  };
+  stream_callbacks.on_cancel_ =
+      [java_stream_callbacks_global_ref](envoy_stream_intel stream_intel,
+                                         envoy_final_stream_intel final_stream_intel) {
+        jvm_on_cancel(stream_intel, final_stream_intel, java_stream_callbacks_global_ref);
+      };
+  stream_callbacks.on_send_window_available_ =
+      [java_stream_callbacks_global_ref](envoy_stream_intel stream_intel) {
+        jvm_on_send_window_available(stream_intel, java_stream_callbacks_global_ref);
+      };
+
   envoy_status_t result = engine->startStream(static_cast<envoy_stream_t>(stream_handle),
-                                              native_callbacks, explicit_flow_control);
+                                              std::move(stream_callbacks), explicit_flow_control);
   if (result != ENVOY_SUCCESS) {
-    env->DeleteGlobalRef(retained_context); // No callbacks are fired and we need to release
+    env->DeleteGlobalRef(
+        java_stream_callbacks_global_ref); // No callbacks are fired and we need to release
   }
   return result;
 }
```

Comment:
the changes in jni_impl.cc _look_ good to me, but I must confess not knowing enough about JNI to do justice to reviewing it

Reply:
The JNI changes are only temporary. I plan to rewrite the entire thing, but that will require changing the Java portion as well, so it's best done in a follow up PR.